File: backend/app/schemas/interview.py
"""Pydantic schemas for Industry interview scheduling
(database/migrations/030_industry_interviews.sql -- the `interviews`
table). Keep field names/constraints in sync with that migration and with
frontend/types/interview.ts.

An interview always hangs off an existing `applications` row (020). Its
`industry_id` and `student_id` are NEVER accepted from the client -- they
are copied server-side (by a database trigger) from the referenced
application, which already carries the authoritative, immutable
student <-> industry <-> posting relationship. `status` is never accepted
from the client either -- it only changes through the explicit lifecycle
endpoints. `extra="forbid"` on the write models is what structurally
rejects an attempt to smuggle any of these in.

Applicant identity: exactly like application responses, an interview
response carries the candidate only as `student_id` (a uuid). The schema
gives Industry no path to an applicant's name/email/profile, and this
module adds none.
"""

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

# database/migrations/030_industry_interviews.sql -- interviews.mode CHECK
InterviewMode = Literal["ONLINE", "PHONE", "ONSITE"]
# database/migrations/030_industry_interviews.sql -- interviews.status CHECK
InterviewStatus = Literal["SCHEDULED", "COMPLETED", "CANCELLED"]

INTERVIEW_MODES: tuple[str, ...] = ("ONLINE", "PHONE", "ONSITE")
INTERVIEW_STATUSES: tuple[str, ...] = ("SCHEDULED", "COMPLETED", "CANCELLED")


class InterviewCreate(BaseModel):
    """POST /api/v1/interviews. Always created as SCHEDULED -- there is no
    `status` field here, and no `industry_id`/`student_id` (derived
    server-side from `application_id`). The application must be one the
    caller owns and must be at the SHORTLISTED or INTERVIEW_SCHEDULED
    stage -- enforced by the service and by a database trigger."""

    model_config = ConfigDict(extra="forbid")

    application_id: str
    scheduled_at: datetime
    duration_minutes: int = Field(default=30, ge=5, le=480)
    mode: InterviewMode
    location: str | None = Field(default=None, max_length=2000)
    notes: str | None = Field(default=None, max_length=10_000)


class InterviewUpdate(BaseModel):
    """PATCH /api/v1/interviews/{id} -- reschedule and/or edit details.
    Partial: only the fields actually sent are changed. Permitted only
    while the interview is still SCHEDULED (the service rejects it once
    COMPLETED/CANCELLED). `application_id` is immutable and cannot be
    changed here; `status` changes only through /complete and /cancel."""

    model_config = ConfigDict(extra="forbid")

    scheduled_at: datetime | None = None
    duration_minutes: int | None = Field(default=None, ge=5, le=480)
    mode: InterviewMode | None = None
    location: str | None = Field(default=None, max_length=2000)
    notes: str | None = Field(default=None, max_length=10_000)


class InterviewOpportunity(BaseModel):
    """The internship/job the underlying application was submitted to.
    `status` here is the posting's lifecycle status
    (DRAFT/PUBLISHED/CLOSED/ARCHIVED), not the interview's."""

    id: str
    title: str
    status: str


class InterviewResponse(BaseModel):
    id: str
    application_id: str
    industry_id: str
    student_id: str
    scheduled_at: str
    duration_minutes: int
    mode: str
    location: str | None = None
    notes: str | None = None
    status: str
    created_at: str | None = None
    updated_at: str | None = None
    # The posting the underlying application was submitted to, resolved
    # through the caller's own postings (RLS permits the join). Mirrors
    # ApplicationResponse.opportunity. Null if the posting embed was
    # unavailable.
    opportunity: InterviewOpportunity | None = None
    opportunity_type: str | None = None


class InterviewListResponse(BaseModel):
    interviews: list[InterviewResponse]

File: frontend/types/interview.ts
// Mirrors the `interviews` table
// (database/migrations/030_industry_interviews.sql) and
// backend/app/schemas/interview.py. Keep all three in sync.
//
// An interview always hangs off an existing `applications` row. Its
// `industry_id` / `student_id` are server-derived from that application
// (never client-supplied), and — exactly like an application response —
// the candidate is only ever exposed as `student_id` (a uuid), never a
// name/email/profile.
//
// Lifecycle: SCHEDULED -> COMPLETED / CANCELLED. Rescheduling is an edit
// of a still-SCHEDULED interview, not a separate status.

import type { OpportunityType } from "@/types/application";

export const INTERVIEW_MODES = ["ONLINE", "PHONE", "ONSITE"] as const;
export type InterviewMode = (typeof INTERVIEW_MODES)[number];

export const INTERVIEW_MODE_LABELS: Record<InterviewMode, string> = {
  ONLINE: "Online",
  PHONE: "Phone",
  ONSITE: "On-site",
};

/** The label for the single `location` field changes by mode (the DB
 * stores a URL for ONLINE, an address for ONSITE, a number for PHONE). */
export const INTERVIEW_LOCATION_LABELS: Record<InterviewMode, string> = {
  ONLINE: "Meeting link",
  PHONE: "Phone number",
  ONSITE: "Address",
};

export const INTERVIEW_LOCATION_PLACEHOLDERS: Record<InterviewMode, string> = {
  ONLINE: "https://meet.example.com/…",
  PHONE: "+91 98765 43210",
  ONSITE: "Office address / room",
};

export const INTERVIEW_STATUSES = ["SCHEDULED", "COMPLETED", "CANCELLED"] as const;
export type InterviewStatus = (typeof INTERVIEW_STATUSES)[number];

export const INTERVIEW_STATUS_LABELS: Record<InterviewStatus, string> = {
  SCHEDULED: "Scheduled",
  COMPLETED: "Completed",
  CANCELLED: "Cancelled",
};

export interface InterviewOpportunity {
  id: string;
  title: string;
  /** The posting's own lifecycle status (DRAFT/PUBLISHED/CLOSED/ARCHIVED). */
  status: string;
}

export interface Interview {
  id: string;
  application_id: string;
  industry_id: string;
  student_id: string;
  /** ISO-8601 UTC instant. Render in the viewer's locale. */
  scheduled_at: string;
  duration_minutes: number;
  mode: InterviewMode;
  location: string | null;
  notes: string | null;
  status: InterviewStatus;
  created_at: string | null;
  updated_at: string | null;
  opportunity: InterviewOpportunity | null;
  opportunity_type: OpportunityType | null;
}

/** POST body — always created as SCHEDULED. No `industry_id`/`student_id`
 * (derived server-side from `application_id`), no `status`. */
export interface InterviewCreate {
  application_id: string;
  scheduled_at: string;
  duration_minutes: number;
  mode: InterviewMode;
  location?: string | null;
  notes?: string | null;
}

/** PATCH body — partial reschedule/edit, SCHEDULED-only. */
export interface InterviewUpdate {
  scheduled_at?: string;
  duration_minutes?: number;
  mode?: InterviewMode;
  location?: string | null;
  notes?: string | null;
}

export const DURATION_OPTIONS = [15, 30, 45, 60, 90, 120] as const;
